runtime/src/main/java/org/corfudb/runtime/clients/ClientResponseHandler.java

```diff
@@ -0,0 +1,235 @@
+                } catch (Throwable e) {
+                    router.completeExceptionally(requestId, e);
+                }
+            }
+        } else if (handlerMap.containsKey(payloadCase)) {
+            try {
+                Object ret = handlerMap.get(payloadCase).handle(response, ctx, router);
+                if (ret != null) {
+                    router.completeRequest(requestId, ret);
+                }
+            } catch (Throwable e) {
+                // TODO: same reason, should we remove this clause?
+                router.completeExceptionally(requestId, e);
+            }
+            return true;
+        }
+        return false;
+    }
+
+    public ClientResponseHandler generateHandlers(@NonNull final MethodHandles.Lookup caller,
+                                                  @NonNull final Object o) {
+        Arrays.stream(o.getClass().getDeclaredMethods())
+                .filter(method -> method.isAnnotationPresent(ResponseHandler.class))
+                .forEach(method -> {
+                    ResponseHandler handler = method.getAnnotation(ResponseHandler.class);
+                    if (!method.getParameterTypes()[0]
+                            .isAssignableFrom(ResponseMsg.class)) {
+                        throw new RuntimeException("Incorrect message type, expected "
+                                + ResponseMsg.class + " but provided "
+                                + method.getParameterTypes()[0]);
+                    }
+                    if (handlerMap.containsKey(handler.type())) {
+                        throw new RuntimeException("Handler for " + handler.type()
+                                + " already registered!");
+                    }
+                    // convert the method into a Java8 Lambda for maximum execution speed...
+                    try {
+                        if (Modifier.isStatic(method.getModifiers())) {
+                            MethodHandle mh = caller.unreflect(method);
+                            MethodType mt = mh.type().changeParameterType(0, ResponseMsg.class);
+                            handlerMap.put(handler.type(), (Handler) LambdaMetafactory
+                                    .metafactory(caller, "handle",
+                                            MethodType.methodType(Handler.class),
+                                            mt, mh, mh.type())
+                                    .getTarget().invokeExact());
+                        } else {
+                            // instance method, so we need to capture the type.
+                            MethodType mt = MethodType
+                                    .methodType(method.getReturnType(), method.getParameterTypes());
+                            MethodHandle mh = caller.findVirtual(o.getClass(), method.getName(), mt);
+                            MethodType mtGeneric = mh.type()
+                                    .changeParameterType(1, ResponseMsg.class)
```

Comment:
You don't need to change the parameter type here. Everything will just be a `ResponseMsg` with nothing extending it. See my other comment above.